primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,441 @@
+    ) -> Vec<F> {
+        if T::size_in_bits() == F::size_in_bits() {
+            RescueSponge::<T, CHUNK_SIZE>::squeeze_native_field_elements_with_sizes(self, sizes)
+                .iter()
+                .map(|x| field_switching(x))
+                .collect::<Vec<F>>()
+        } else {
+            unimplemented!("currently we do not support hashing into a non-native field");
+        }
+    }
+
+    /// Squeeze `num_elements` nonnative field elements from the sponge.
+    ///
+    /// Because of rust limitation, for field-based implementation, using this
+    /// method to squeeze native field elements will have runtime casting
+    /// cost. For better efficiency, use `squeeze_native_field_elements`.
+    fn squeeze_field_elements<F: PrimeField>(&mut self, num_elements: usize) -> Vec<F> {
+        self.squeeze_field_elements_with_sizes::<F>(
+            vec![FieldElementSize::Full; num_elements].as_slice(),
+        )
+    }
+
+    /// Creates a new sponge with applied domain separation.
+    fn fork(&self, domain: &[u8]) -> Self {
+        let mut new_sponge = self.clone();
+
+        let mut input = Absorb::to_sponge_bytes_as_vec(&domain.len());
+        input.extend_from_slice(domain);
+        new_sponge.absorb(&input);
+
+        new_sponge
+    }
+}
+
+/// The interface for field-based cryptographic sponge.
+/// `CF` is the native field used by the cryptographic sponge implementation.
+impl<T: RescueParameter, const CHUNK_SIZE: usize> FieldBasedCryptographicSponge<T>
+    for RescueSponge<T, CHUNK_SIZE>
+{
+    /// Squeeze `num_elements` field elements from the sponge.
+    fn squeeze_native_field_elements(&mut self, num_elements: usize) -> Vec<T> {
+        // SQUEEZE PHASE
+        let permutation = Permutation::default();
+        let mut result = vec![];
+        let mut remaining = num_elements;
+        // extract current rate before calling PRP again
+        loop {
+            let extract = remaining.min(RATE);
+            result.extend_from_slice(&self.state.vec[0..extract]);
+            remaining -= extract;
+            if remaining == 0 {
+                break;
+            }
+            self.state = permutation.eval(&self.state)
+        }
+        result
+    }
+
+    /// Squeeze `sizes.len()` field elements from the sponge, where the `i`-th
+    /// element of the output has size `sizes[i]`.
+    fn squeeze_native_field_elements_with_sizes(&mut self, sizes: &[FieldElementSize]) -> Vec<T> {
+        let mut all_full_sizes = true;
+        for size in sizes {
+            if *size != FieldElementSize::Full {
+                all_full_sizes = false;
+                break;
+            }
+        }
+
+        if all_full_sizes {
+            self.squeeze_native_field_elements(sizes.len())
+        } else {
+            // we do not currently want to output field elements other than T.
+            // This will be fixed once `squeeze_bytes` interfaces is fixed.
+            unimplemented!()
+        }
```

Comment:
again, fill unimplemented with reason string.

Reply:
Done, unified all the unimplemented messages to: "Currently we only support squeezing native field elements!" in https://github.com/EspressoSystems/jellyfish/pull/146/commits/995ec471d37e5b3ef55cbdacee127cdd2c8afabd.